cdap-app-fabric/src/main/java/co/cask/cdap/security/DefaultUGIProvider.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright © 2016 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.security;
+
+import co.cask.cdap.common.conf.CConfiguration;
+import co.cask.cdap.common.conf.Constants;
+import co.cask.cdap.common.kerberos.SecurityUtil;
+import co.cask.cdap.common.utils.DirUtils;
+import co.cask.cdap.data2.security.ImpersonationInfo;
+import co.cask.cdap.data2.security.ImpersonationUtils;
+import co.cask.cdap.data2.security.UGIProvider;
+import com.google.inject.Inject;
+import org.apache.hadoop.security.UserGroupInformation;
+import org.apache.twill.filesystem.LocationFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.URI;
+
+/**
+ * Provides a UGI by logging in with a keytab file for that user.
+ */
+public class DefaultUGIProvider implements UGIProvider {
+
+  private static final Logger LOG = LoggerFactory.getLogger(DefaultUGIProvider.class);
+
+  private final CConfiguration cConf;
+  private final LocationFactory locationFactory;
+
+  @Inject
+  public DefaultUGIProvider(CConfiguration cConf, LocationFactory locationFactory) {
+    this.cConf = cConf;
+    this.locationFactory = locationFactory;
+  }
+
+  /**
+   * Resolves the {@link UserGroupInformation} for a given user, performing any keytab localization, if necessary.
+   *
+   * @return a {@link UserGroupInformation}, based upon the information configured for a particular user
+   * @throws IOException if there was any IOException during localization of the keytab
+   */
+  @Override
+  public UserGroupInformation getConfiguredUGI(ImpersonationInfo impersonationInfo) throws IOException {
+    LOG.debug("Configured impersonation info: {}", impersonationInfo);
+    File tempDir = DirUtils.createTempDir(new File(cConf.get(Constants.CFG_LOCAL_DATA_DIR),
```

Comment:
You might want to use `createTempDirectory(Path dir, String prefix, FileAttribute<?>... attrs)` so that the directory created is only readable by this user (cdap). Also, prior to that call, you'll need to create the "data/temp" directory first as it might not exists. Usually I just chain them up.

`Files.createTempDirectory(Files.createDirectory(Paths.get(cConf.get(...), cConf.get(...)), ...)`